File: Matlab Files/Beam_Fitting.m
%% Gaussian Beam Fitting
%%%For list of 1/e^2 sizes and positions, fit to beam propagation equation
%%%to get Rayleigh range, waist position, and waist size of beam
%%%Future additions:  fitting in both x and y transverse dimensions

close all
clear all

%% Initialization
fontsize = 16;
lambda = 690e-9; %[m] wavelength of the laser beam
errorflag = 1; %1 to include error bars as weights in fitting; 0 otherwise

%% Import data
%%%Data should be in 2 column format: longitudinal position (z) and 1/e^2
%%%radius of the beam at that longitudinal position
filepath = char('/Users/work/Documents/Nanofibers/Laser Profiling/Blue-detuned Laser Profiling/Telescope 2 Output Side/');
inputfilename = char('ProbeLaser_NanofiberOutput_Telescope1_20130906_redcoord.txt');
filename = strcat(filepath,inputfilename);
[z beamsize]=textread(filename,'%f%f','commentstyle','matlab'); %read in data file, z in cm and beamsize in mm

%% Data preparation
z = z.*1e-2; %[m] converts z from cm to m
beamsize = beamsize.*1e-3; %[m] converts beamsize from mm to m
lambdavector = zeros(1,length(z));
lambdavector(1) = lambda;

%% Fitting
%%%Fit to beam propagation equation
InitialGuess=[1e-4 3]; %[m m] initial guess for waist value and position of waist
datamatrix = [z(:) lambdavector(:)]; %[m m] data matrix contains both the values to be fitted and the fixed parameter (laser wavelength)
if errorflag==1
    errorvector = beamsize./beamsize; %weight all points equally in absence of better information
    %errorvector=[0.8 0.4 0.4 0.2 0.1];
    [P,r,J]=nlinfitweight(datamatrix,beamsize(:),@fittogaussianbeampropagation,InitialGuess,errorvector(:));
else
    [P,r,J]=nlinfit(datamatrix,beamsize(:),@fittogaussianbeampropagation,InitialGuess);
end

fitwaist = P(1) %[m] waist size; minimum beam size
fitwaistposition = P(2) %[m] longitudinal position of minimum beam size

%% Generate vector that represents the fit to the points
stepsize = (max(z)-min(z))/100;
fitz=-3:.001:3; %fitz = min(z):stepsize:max(z);
zR = pi*fitwaist^2/lambda; %[m] Rayleigh range of the beam
fitbeamsize = fitwaist.*sqrt(1+((fitz-fitwaistposition)./zR).^2);%[m] 1/e^2 beam size as a function of z position

%% Plot the data, the fit, and the fit parameters
figure(1)
if errorflag==1
    errorbar(z(:).*1e2,beamsize(:).*1e3,errorvector(:)./10,'sr','MarkerFaceColor','r','MarkerSize',8)
else
    plot(z(:).*1e2,beamsize(:).*1e3,'sr','MarkerFaceColor','r','MarkerSize',8)
end
hold on
plot(fitz(:).*1e2,fitbeamsize(:).*1e3,'-k','LineWidth',1.2)
%ylim([0 (max(beamsize(:))+0.1.*max(beamsize(:))).*1e3]);
ylim([0 1.9]);
%xlim([50 150]);
set(gca,'FontSize',fontsize,'FontWeight','bold');
xlabel('Longitudinal Position [cm]','FontSize',fontsize,'FontWeight','bold');
ylabel('1/e^2 Beam Radius [mm]','FontSize',fontsize,'FontWeight','bold');
text(min(fitz.*1e2),max(fitbeamsize.*1e3),strcat('w0 = ',num2str(fitwaist.*1e3,3),' mm'),'FontSize',fontsize,'FontWeight','bold');
text(min(fitz.*1e2),max(fitbeamsize.*1e3)-max(fitbeamsize.*1e3)./50,strcat('z0 = ',num2str(fitwaistposition.*1e2,3),' cm'),'FontSize',fontsize,'FontWeight','bold');
%text(min(fitz.*1e2),1.2,strcat('w_0 = ',num2str(fitwaist.*1e3,3),' mm'),'FontSize',fontsize,'FontWeight','bold');
%text(min(fitz.*1e2),1,strcat('z_0 = ',num2str(fitwaistposition.*1e2,3),' cm'),'FontSize',fontsize,'FontWeight','bold');
%title('894 nm input to Nanofiber','FontSize',fontsize,'FontWeight','bold');

File: Python Conversions/Beam_Fitting.py
## Gaussian Beam Fitting
# For list of 1/e^2 sizes and positions, fit to beam propagation equation to get Rayleigh range, waist position, and waist size of beam
# Future additions: simultaneous fitting in both x and y transverse dimensions

import numpy as np                # Import numpy functions
import scipy.optimize as optimize # Import optimize function for least squares fitting
import matplotlib.pyplot as plt          # Import matplotlib functions

plt.close()

## Initialization
fontsize = 16         # Font size for plotting
laserlambda = 894e-9  #[m] wavelength of the laser beam
errorflag = 0         #1 to include error bars as weights in fitting; 0 otherwise

## Import data
# Data should be in 2-column format: longitudinal position (z) and 1/e^2 radius of the beam at that longitudinal position
directoryName = '/Users/work/Documents/Arizona/Nanofibers/Laser Profiling/Probe Beam Profiling/Before Fiber Link/'
inputFileName = 'ProbeLaser_FiberLink_ReverseOutput_20131008.txt'
fileName = directoryName + inputFileName
importedData = np.loadtxt(fileName)

zData = importedData[0]         #[cm] Longitudinal position along laser beam
beamSize = importedData[1]      #[mm] Transverse position across laser beam

## Data preparation
zData = zData*1e-2              #[m] converts z from cm to m
beamSize = beamSize*1e-3        #[m] converts beamSize from mm to m

## Fitting
#Fit to beam propagation equation
def GaussianPropagate(z,p):
  return p[0] * np.sqrt(1 + ( (z - p[1]) * laserlambda / (np.pi * p[0]**2) )**2)
  
def residuals(p,y,z):
  return y - GaussianPropagate(z,p)

waistGuess = 0.1e-3     #[m] Initial guess for waist value
zGuess = -0.7          #[m] Initial guess for position of waist

plsq = optimize.leastsq(residuals,[waistGuess,zGuess],args = (beamSize,zData))
fitWaist = plsq[0][0]         #[m] waist size; minimum beam size
fitWaistPosition = plsq[0][1] #[m] longitudinal position of minimum beam size

zFit = np.arange(-1,2,0.001)
beamPropagate = GaussianPropagate(zFit,[fitWaist,fitWaistPosition])#+0.00065
#beamPropagateGuess = GaussianPropagate(zFit,[0.35e-3,0.7])+0.00145

## Plot waist size versus distance from lens
plt.figure(1)
plt.plot(zData,beamSize,'sr',label="Measurements")
plt.plot(zFit,beamPropagate,'--b',label="Best Fit")
#plt.plot(zFit,beamPropagateGuess,'--g')
plt.xlabel('Longitudinal Position [m]')
plt.ylabel('Waist Size [mm]')
plt.text(0.05*np.mean(zData),1.1*np.mean(beamPropagate),'1/$e^2$ radius = '+ str(round(fitWaist,5)) + ' m')
plt.text(0.05*np.mean(zData),1.05*np.mean(beamPropagate),'z$_0$ = '+ str(round(fitWaistPosition,4)) + ' m')
plt.legend(bbox_to_anchor=(0, 0, 1, 1), bbox_transform=plt.gcf().transFigure)
plt.title('Beam Profile of Nd:YAG Laser')
